### ISSUE TITLE
Don’t Let LGTMs Become Stale for Branch Updates

### ISSUE THREAD

**Issue by jGleitz** [Reviewable/Reviewable#226]
To keep ourselves from accidentally merging bad code, we use required status checks on GitHub. As a logical consequence, we have to update our branches regulary (after every accepted PR). We require 2 lgtms on Reviewable.

Now in phases were we get a lot of stuff done, we often experience the issue that a PR gets 2 lgtms but needs to be updated. When hitting the “update” button on GitHub (=> there where no conflicts), no new information is added that could be reviewed. Still, the lgtms become stale and need to be updated. This seriously slows down our workflow, as 3 (!) people need to take action to merge an *already accepted* PR (author needs to update, two reviewers need to update their lgtm). We do not want to use :lgtm_strong: to avoid minor changes not being reviewed.

To solve this, I see two possibilities:

**The suboptimal**: Offer the commit messages in the data for the review completion code. This would enable us to ignore commits with the message `Merge branch 'master' into 'pr-source'`.

**The optimal**: Offer an “update branch” button on Reviewable. At the moment, Reviewable often lets one try to merge a PR even though it is out-of-date and hence shows an error message. If the PR author could hit “update branch” on Reviewable, that would make stuff easier. But, more importantly: Reviewable would me able to mark the merge commit as “commit that did not introduce changes” (because the review was against `master` in the first place) and not let the lgtms become stale for these special commits. This would target exactly our use case and would not allow developers to sneak in changes by choosing a clever name for the commit message.

**Comment by pkaminski** [comment]
Thanks for the detailed issue writeup!  Let me make sure I understand it correctly before we start talking about potential solutions, though:

> As a logical consequence, we have to update our branches regulary (after every accepted PR).

To be completely clear:  you only accept fast-forward merges into master to make sure that a non-FF merge (even if clean) doesn't introduce a bug, right?  So you need to update your feature branches whenever master changes (via rebasing or back-merging from master) and test the new state to ensure that it's working?

> When hitting the “update” button on GitHub (=> there where no conflicts)

What do you mean by this?  I don't remember seeing such a button in my PRs, and AFAIK status checks generally update automatically whenever you push to the PR's source or target branches.  Is this some way to automatically merge from the target branch back into the source branch of a PR?

> no new information is added that could be reviewed.

Do you mean that Reviewable classifies all the deltas between the revisions as "base" and hides them by default?  If you have some handy public PRs, could you point to a sample review where this situation occurred and say which revisions correspond to which points in your scenario?

> Offer an “update branch” button on Reviewable. At the moment, Reviewable often lets one try to merge a PR even though it is out-of-date and hence shows an error message.

Reviewable should automatically update the mergeability status of a PR; normally within seconds of a push, and within 10-60 seconds for GitHub's own mergeability flag (which requires polling).  If you're seeing latencies higher than that can you describe the scenario in more detail?  Thanks.

**Comment by jGleitz** [comment]
I fear that I fail to exactly describe what I mean because of wrong use of vocabulary (I’m neither a native english speaker nor a git-master). I will include several screenshots to show what I am talking about:

> To be completely clear: you only accept fast-forward merges into master to make sure that a non-FF merge (even if clean) doesn't introduce a bug, right? So you need to update your feature branches whenever master changes (via rebasing or back-merging from master) and test the new state to ensure that it's working?

Essentially, that’s correct. We have mostly PRs from private forks and not from feature branches, but that does not change much.

We enforce this policy by using the GitHub “protect branch” feature:

![master-branch-protection](https://cloud.githubusercontent.com/assets/4305652/12141663/c589e720-b471-11e5-8cf7-7f15b57d5b88.png)

Because of this, successful automatic testing on Travis and a complete Review on Reviewable are necessary before we are technically able to click the “merge” button:

![pending-checks](https://cloud.githubusercontent.com/assets/4305652/12141695/099972c8-b472-11e5-9f4d-187a4ce2b328.png)



> >When hitting the “update” button on GitHub (=> there where no conflicts)

> What do you mean by this? I don't remember seeing such a button in my PRs, and AFAIK status checks generally update automatically whenever you push to the PR's source or target branches. Is this some way to automatically merge from the target branch back into the source branch of a PR?

This button only shows up if the  “protect branch” feature is enabled for the target branch. And yes, it automatically merges the target branch into the source branch.

This feature makes a lot of sense for Travis: There could for examble be two PRs. Number 1 adds additional tests on Travis, Number 2 adds some code. Both pass Reviewable and Travis. But Number 1 introduces a Test that Number 2 does not pass. So as soon as Number 1 gets merged into `master`, GitHub shows that Number 2 is out-of-date: It has to be merged with the latest changes, to be tested again on Travis, together with the changes of Number 1:

![out-of-date](https://cloud.githubusercontent.com/assets/4305652/12141796/fd410148-b472-11e5-8fff-25ee529a948d.png)

As a convenience feature, GitHub offers the “update branch” button to the owners of the source branch. In times when up to 10 PRs get merged before another one (reality for us), it’s nice that you can simply hit it instead of merging 10 times on your local machine:

![update-branch](https://cloud.githubusercontent.com/assets/4305652/12141926/2a45cf56-b474-11e5-986a-a0ec68295d05.png)




> > no new information is added that could be reviewed.

> Do you mean that Reviewable classifies all the deltas between the revisions as "base" and hides them by default? If you have some handy public PRs, could you point to a sample review where this situation occurred and say which revisions correspond to which points in your scenario?

Yes, exactly. To stick to the above example: As soon as the owner of Number 2 hits “update branch”, all lgtms are lost. But there is nothing new to review. That makes a lot of sense, if you think about it: When comparing to `master`, merging with `master` of course does not introduce any changes.

We for example had this situation for Beagle-PSE/Beagle#362. `r3`-`r9` are only such merge commits created by the “update branch” button. We’d have to renew our :lgtm: after each revision:

![nothing-to-review](https://cloud.githubusercontent.com/assets/4305652/12142060/6d073e00-b475-11e5-8411-5beed4cabc8c.png)


I realise that my proposed solutions have downsides. So I’ll only describe what we’d like to have:

In the example above, after giving a :lgtm:, we don’t want to have it revoked for merge commits that do not introduce any changes when comparing against the PR’s target branch. But opposed to :lgtm_strong:, we want to have it revoked as soon as any commit adds a new (not yet reviewed) change.

I hope I managed to make clear what I’m talking about. Please do not hesitate to ask if there is any information we can provide to help with this issue!

**Comment by pkaminski** [comment]
Thank you for the extra details, and especially the link to the sample review!  I think I now understand what's going on:  there's a flaw in Reviewable's revision partitioning algorithm where it creates empty revisions even though there are absolutely no changes in the PR's files.  (That is, it's not that there are base changes that are hidden by default -- there's really no changes at all.)  This is my fault for not foreseeing such "empty" commits.  I'm working on a fix right now and should have it ready in a day or two, once it's thoroughly tested.  When it's deployed, empty revisions should no longer be created (though previously created ones will remain), and your LGTMs should work as expected.

Sorry for the trouble!  And BTW, your English and vocabulary are perfectly fine -- it's my lack of familiarity with this workflow that's the problem.  :smile:

**Comment by pkaminski** [comment]
I update the revision mapping algorithm so "empty" revisions should no longer appear.  It's not retroactive, though, and the code is a bit tricky so while it works in my tests there may still be corner cases.  Please reopen the issue (ideally with a link to the misbehaving review) if it's not quite working right.
